awscli/testutils.py

```diff
@@ -786,10 +789,20 @@ def put_object(self, bucket_name, key_name, contents='', extra_args=None):
         response = client.put_object(**call_args)
         self.addCleanup(self.delete_key, bucket_name, key_name)
 
-    def delete_bucket(self, bucket_name):
+    def delete_bucket(self, bucket_name, attempts=5, delay=5):
         self.remove_all_objects(bucket_name)
         client = self.create_client_for_bucket(bucket_name)
-        response = client.delete_bucket(Bucket=bucket_name)
+
+        # There's a chance that, even though the bucket has been used
+        # several times, the delete will fail due to eventual consistency
+        # issues.
+        for _ in range(attempts):
+            try:
+                client.delete_bucket(Bucket=bucket_name)
+                break
+            except client.exceptions.NoSuchBucket:
+                time.sleep(delay)
```

Comment:
This has the downside that we won't fast fail on invalid bucket names, or in the edge case where the bucket is actually deleted but we get a 404.  The step (3) that I suggested would allow us to short circuit this condition, with the caveat that HEAD bucket and DELETE bucket have to agree.

Reply:
I don't think that's something we really have to worry about. There's no reason for us to be calling delete on a bucket multiple times. Introducing that condition re-introduces the risk of the different paths not having correct information.

Reply:
I still think is worth having a fast fail in the case of errors, even if it just happens when a dev is working on a new test, shorter feedback loops really help.  A 25 second wait penalty seems unfortunate.

Like I mentioned, the *only* issue here is that both delete_bucket() and head_bucket() say the bucket doesn't exist, and yet the bucket still somehow exists.  In that case the test won't fail but we'll leak a bucket in our test runs.

If we were actually worried about that case we could add a similar `min_success` thing with our head_bucket calls.